improved ind2sub/sub2ind

Jutho opened the pull request:
Use `stagedfunction` definition of `sub2ind` and `ind2sub` to generate specialized code for every `N` in `dims::NTuple{N,Integer}`. See JuliaLang/julia#10331 for related discussion and performance check. I can do more performance benchmarking if somebody can point me to a good test suite that depends heavily on the use of those 2 functions.

ViralBShah commented:
I wonder if there are better names for these functions, that can be significantly better than the matlab names.

ViralBShah commented:
BTW, what are the performance gains on some examples? I do not have a code handy that uses these a lot.

ViralBShah commented:
Sparse indexing and various `find*` routines on sparse matrices use these functions. Perhaps this helps some sparse perf benchmarks?

Jutho commented:
I also don't have a set of benchmark examples and the problem is that they get optimized away in a simple for loop that just tries to call this functions a number of times with fixed argument. Ah but maybe I can avoid that behavior by letting the arguments of the function call depend on the iteration variable. I'll report back soon.

Jutho commented:
With the following test code (only for the hard case: `ind2sub`)
```julia
function test(N)
    d1=N
    @time for i=1:N;s1=ind2sub((d1,),i);end
    d2=ceil(Int,N^(1/2))
    @time for i=1:N;s2=ind2sub((d2,d2),i);end
    d3=ceil(Int,N^(1/3))
    @time for i=1:N;s3=ind2sub((d3,d3,d3),i);end
    d4=ceil(Int,N^(1/4))
    @time for i=1:N;s4=ind2sub((d4,d4,d4,d4),i);end
    d5=ceil(Int,N^(1/5))
    @time for i=1:N;s5=ind2sub((d5,d5,d5,d5,d5),i);end
end
```
I obtain on master:
```
elapsed time: 3.38e-7 seconds (0 bytes allocated)
elapsed time: 0.001355097 seconds (0 bytes allocated)
elapsed time: 0.02040562 seconds (0 bytes allocated)
elapsed time: 0.290835901 seconds (228 MB allocated, 4.73% gc time in 10 pauses with 0 full sweep)
elapsed time: 0.382516052 seconds (305 MB allocated, 4.53% gc time in 14 pauses with 0 full sweep)
```
and with this PR:
```
elapsed time: 4.26e-7 seconds (0 bytes allocated)
elapsed time: 0.001568929 seconds (0 bytes allocated)
elapsed time: 0.043808195 seconds (0 bytes allocated)
elapsed time: 0.059630003 seconds (0 bytes allocated)
elapsed time: 0.082583044 seconds (0 bytes allocated)
```
Clearly, the 1D case is just optimized away (the time does not depend on `N`) both on master and with this PR.  In this PR, the code for the 1d case is just `ind-=1;return (ind+1,)`. I guess the `-1 +1` will always be optimized away, or will never have any significant effect`.

The 2d case is roughly the same or slightly slower. The 3d case seems a factor 2 slower with this PR, this I will need to fix, either by a specialized version or a more clever formulation of the `stagedfunction`. I could also add a specialized version for the 1d case to get rid of the `-1+1` and anyway noticed that I need to fix two ambiguity warnings in the build process.

For higher dimensions, this PR is significantly faster.

Jutho commented:
Not sure what was wrong before, but the new timings with the latest commit are:
```
elapsed time: 2.81e-7 seconds (0 bytes allocated)
elapsed time: 0.001192448 seconds (0 bytes allocated)
elapsed time: 0.010499385 seconds (0 bytes allocated)
elapsed time: 0.019768971 seconds (0 bytes allocated)
elapsed time: 0.031037068 seconds (0 bytes allocated)
```
which beat the timings on master for every case, and even more significantly as before in the higher-dimensional case.

Jutho commented:
grepping `ind2sub` in `julia/base/*.jl` only returns function definitions, so it seems like this is not used anywhere in `find`. There are indeed some occurrences in `julia/base/sparse/sparsematrix.jl` but these are all the two-dimensional case, so I don't think there will be any performance effect.

Jutho commented:
The improved timings seem to come from the fact that I now explicitly forced `ind2sub` to be inlined. I forgot that I added that line, but it seems to have a significant effect (for this particular test function). Not sure if it is always a good idea to inline this, although the generated code is pretty short ( `2N+1` lines for `N` dimensions).

JeffBezanson commented:
We really have to avoid using staged functions for too many things. They make it basically impossible to statically compile programs. Staged functions become totally useless with the JIT off. For simpler functions like this, I think we should try to do compiler improvements and rewrite them in other ways.

kmsquire commented:
For the purpose of statically compiling, wouldn't it be possible to keep track of what functions were generated by staged functions, e.g., during a particular run of code, and allowing the user to precompile those functions?  Obviously, this wouldn't work if the final code took a different code path (although even here, an appropriate error could be issued), but a 90% solution could still be quite useful.

JeffBezanson commented:
> Obviously, this wouldn't work if the final code took a different code path

Exactly :)

Jutho commented:
> We really have to avoid using staged functions for too many things. They make it basically impossible to statically compile programs. Staged functions become totally useless with the JIT off. For simpler functions like this, I think we should try to do compiler improvements and rewrite them in other ways.

I didn't know this. I am happy with closing this PR, or trying to see whether I can improve the `ind2sub/sub2ind` code without `stagedfunction`s, but I don't really know what the general conclusion is. What should be the general rule when one can use `stagedfunction`? Would it help if e.g. the most used cases (e.g. 1 and 2 dimensions, possibly up to 3 or 4) are regular method definitions and then the higher ones are a stagedfunction or is this also not ok.

More generally, to avoid having to write `stagedfunction` definitions in the case where I use them often, it would help if there would be the following functionality:
* an `ndims(a::AbstractArray)` method that returns a result where the dimensionality is encoded as a type parameter, e.g. something like `Val{N}` but then maybe a dedicated type (e.g. `Dim{N}`?) that supports some basic arithmetic (`min`, `max`, `+`, `-`, ...)
* a function to create a tuple where the length and the type of values can be inferred, e.g. something like `ntuple(::Dim{N}(), sometypedfunction)`.

Is this worth discussing in a separate issue?

JeffBezanson commented:
Yes, it would be great to develop more tricks for writing code like this without the full sledgehammer of staged functions, e.g. #5560.

Generally what works now is "unary" encodings of numbers in tuple lengths, e.g. https://gist.github.com/JeffBezanson/10612461.

Jutho commented:
Thanks. I think I once saw that Gist once before, but had forgotten about it. This is neat.

I'd still like some specific advice on how to precede with this PR. I still find it strange that these functions (`ind2sub/sub2ind`), which have hardly any use within `Base`, should not be `stagedfunction`s, whereas several of the more basic functions for multidimensional arrays that have recently been rewritten all use or depend on `stagedfunction`.

tknopp commented:
I use `ind2sub` in https://github.com/tknopp/NFFT.jl for implementing a slow reference implementation for the NDFT. There is also a test file.
When I wrote this, there was no 1D fallback for `ind2sub`. Not sure if this is included in this PR.

Jutho commented:
Ok, using `tail` on the `dims` argument I was able to write non-staged versions of these functions which improve on the existing definitions. With the test function above (but also including an `N=6` case), master yields
```
test(1000000)
elapsed time: 3.65e-7 seconds (0 bytes allocated)
elapsed time: 0.001170862 seconds (0 bytes allocated)
elapsed time: 0.019838434 seconds (0 bytes allocated)
elapsed time: 0.294538723 seconds (228 MB allocated, 4.08% gc time in 10 pauses with 0 full sweep)
elapsed time: 0.372374532 seconds (305 MB allocated, 3.53% gc time in 14 pauses with 0 full sweep)
elapsed time: 0.460680693 seconds (389 MB allocated, 5.09% gc time in 18 pauses with 0 full sweep)
```
and this latest PR yields:
```
julia> test(1000000)
elapsed time: 2.8e-7 seconds (0 bytes allocated)
elapsed time: 0.001197907 seconds (0 bytes allocated)
elapsed time: 0.010452432 seconds (0 bytes allocated)
elapsed time: 0.060237796 seconds (0 bytes allocated)
elapsed time: 0.157890394 seconds (122 MB allocated, 2.92% gc time in 6 pauses with 0 full sweep)
elapsed time: 0.24761047 seconds (205 MB allocated, 5.06% gc time in 9 pauses with 0 full sweep)
```
So this PR is clearly faster, but has allocation overhead as soon as `N>4`, which will hopefully go away with better compilation. In particular, there is already a significant difference, since unlike with the definitions in master, the return type is correctly inferred also for `N>= 4`, and so this allocation overhead is not due to the type not being known.

timholy commented:
@JeffBezanson, to clarify: a statically-compiled program would still have type inference->native code generation, but not parsing/lowering? Otherwise I don't understand the distinction you're making, and would like to know more.

timholy commented:
(The point being that
> Obviously, this wouldn't work if the final code took a different code path

is relevant even for generic functions that have not yet been JITed.)

Jutho commented:
> is relevant even for generic functions that have not yet been JITed.)

I was wondering about the same thing but didn't dare to ask as I really don't know how these things work deep down.

jakebolewski commented:
You would have to fallback on running code through an interpreter.

JeffBezanson commented:
That is correct. Then one problem is that we don't have an interpreter for the full language; e.g. llvm intrinsics are not supported. The other problem is that it would be extremely slow, especially since staged functions are typically used for performance.

timholy commented:
Sorry to be dense, but let me ask for a direct answer to my real question: is JITing still available in a statically-compiled program? If not, then I still don't see any difference between staged functions and generic functions.

timholy commented:
Oh, I think I now understand which part of my question @jakebolewski was responding to, and so my understanding is:
- JIT is not supported in a statically-compiled program
- One could run code with an interpreter
- Currently interpreter support is limited

If this is right, then the concern about staged functions still seems to be a red herring: if you cache the code returned by the stagedfunction, by design it's going to be faster to run via the interpreter than whatever code the stagedfunction would have replaced---the whole point of a stagedfunction is to generate optimized julia code.

StefanKarpinski commented:
I believe that part of the confusion stems from there being two versions of "statically compiled": with and without JIT support. With JIT support, staged functions are no problem. Without JIT support, they are a problem. The difference between staged functions and macros in this respect is that you can know that there are no more macro expansions required, whereas it doesn't seem possible to know that you won't encounter more code generated by a staged function. For normal generic functions, you can always generate a slow, generic fallback, but for staged functions, you can't even do that because you don't yet have the code that you need to generate the slow generic fallback for.

mbauman commented:
> For normal generic functions, you can always generate a slow, generic fallback, but for staged functions, you can't even do that because you don't yet have the code that you need to generate the slow generic fallback for.

Doesn't it just require two slow interpreted passes?  One to generate the staged AST and a second to interpret it?  There'd need to be some work to cache AST instead of the actual function, of course, but I don't see why it's that different.

Jutho commented:
Otherwise, might it be an option to requie a non-staged fallback for every stagedfunction, which in the worst case just trows a `MethodError` but preferably contains a less efficient but still working alternative.

JeffBezanson commented:
Yes, technically an interpreter can be used (as already discussed above).

However there is another reason to discourage staged functions, which is that their type inference is basically all-or-nothing, to preserve type safety. This is a tough problem. We could, for example, store all results and check monotonicity directly, giving an error if we catch a staged function "lying". But this is pretty complicated, and also pretty hard on staged function authors. It also leaves you open to unanticipated errors far in the future.

Jutho commented:
For a non-specialist, what do you mean with the monotonicity argument?

And a completely different question: Even though I am enjoying the general discussion and learning a lot, I'd also like to know the specific opinion on the current PR? I believe even the non-staged version in the latest commit still provides an improvement over the original definitions, though the simple benchmark results are of course less convincing then those of the `stagedfunction` approach used initially. The major improvement over the original code is that the recursive definition allows to infer the correct return type of `ind2sub` for higher `N`.

JeffBezanson commented:
I could write a devious staged function like this:

```
stagedfunction wrong(x)
    if x === Any
        "str"
    else
        1
    end
end
```

On all actual arguments, this function returns `1`, but if it's called on `Any` during type inference, it "claims" to return a string. In this extreme case we therefore infer an *incorrect* type, which can lead to segfaults and so on. Unfortunately it is relatively easy to write staged functions that have this problem.

Jutho commented:
The `ind2sub` definitions could be even improved by forcing them to be inlined and replacing the `rem` instruction with a multiplication using the result of the `div` calculation. This leads to the following definitions
```julia
@inline ind2sub(dims::(Integer,Integer), ind::Integer) = begin
	ind2 = div(ind-1,dims[1])+1
	(ind - dims[1]*(ind2-1), ind2)
end
@inline ind2sub(dims::(Integer,Integer,Integer...), ind::Integer) = begin
	ind2 = div(ind-1,dims[1])+1
    tuple(ind-dims[1]*(ind2-1), ind2sub(Base.tail(dims),ind2)...)
end
```
and the following timing results
```
elapsed time: 2.5e-7 seconds (0 bytes allocated)
elapsed time: 0.007854144 seconds (0 bytes allocated)
elapsed time: 0.087311776 seconds (0 bytes allocated)
elapsed time: 0.162448165 seconds (0 bytes allocated)
elapsed time: 0.258635037 seconds (0 bytes allocated)
elapsed time: 0.369751561 seconds (0 bytes allocated)
```
Note that all allocation overhead is gone. However, since `@inline` cannot yet be used in `abstractarray.jl`, I would need to replace these definitions to elsewhere. I can wrap this in a proper commit over the weekend.

timholy commented:
Very nice, although there's still a large gap between this and your best stagedfunction implementation. To me it's not obvious one couldn't match it: is there any chance one could do even better with a two-pass algorithm? The idea might be to first move tailward to compute strides, and second move headward computing the subscripts? That's closer to how one might write it out in cases of fixed dimensionality.

Jutho commented:
There was something wrong with those timings, I ran the for loop ten times more then before, so they are roughly of by a factor of time. I updated the timings in the previous post

Jutho commented:
Which is actually much faster than my initial `stagedfunction` results. Also, my timings are not consistent since some are on my laptop and some on my desktop. I will try a more serious benchmark later this weekend.

timholy commented:
Ah, much better. Thanks.

carlobaldassi commented:
> However, since `@inline` cannot yet be used in `abstractarray.jl`

FWIW I had done a crude implementation here: https://github.com/JuliaLang/julia/commit/46087b93c71fd2e4d3f7128310a6dd0f816dcf92

Jutho commented:
Thanks for the tip. If #9607 gets merged I can use the `_crude_meta`, otherwise I will just move the definitions to somewhere where I can use the `@inline` macro.

timholy commented:
@Jutho, when you get around to this, I wonder if we should deprecate [this definition](https://github.com/JuliaLang/julia/blob/f584992b3edafd6041870f8775b75c78a722a30e/base/abstractarray.jl#L1052-L1066) (or change the second input to a tuple), and force sizes to always be specified by tuples. The reasoning is to clear space for new functionality:
- given https://github.com/JuliaLang/julia/issues/8188#issuecomment-56763806, it wouldn't be surprising to see `AbstractArray` types crop up that store magic numbers for fast linear indexing
- when the magic numbers are available, being able to use `ind2sub(A::AbstractArray, ind::Integer)` or even `ind2sub(A::AbstractArray, ind::Integer, dim::Integer = 1)` (the latter assuming that `ind` applies starting at dimension `dim` of `A`) would be a major performance advantage.
- to make that available, we'd better remove any potential confusion with array-as-a-list-of-dimension-sizes.

There's no actual conflict now, but that one method is a bit inconsistent with the rest. (I have no problem with the `sub` _output_ of that function, just the `dims` input.)

garrison commented:
While we're on the subject of `ind2sub`, I've always found it strange that the tuples returned are not ordered the same as the indices given (according to `<`).  For instance,

    julia> a, b = 3, 4

    julia> dims = (3,6)

    julia> a < b
    true

    julia> ind2sub(dims, a) < ind2sub(dims, b)
    false

To "fix" this would mean reversing the tuple returned, such that the last index is incremented first.  I'm left wondering why the current behavior was chosen.  I can only guess it's to be consistent with Matlab, though it also means that the ordering of column-major matrix indices is respected.  Still, having the returned tuples ordered wrong seems like a wart to me.

timholy commented:
I think `ind2sub` is used mostly in the context of array indexing. Admittedly, it's a much less-needed function in Julia than in Matlab because we have much more sophisticated iterators, but I doubt we want to change the order. One option would be to create a variant `ind2sub(::ReverseOrdering, ...)` that would do what you suggest.

GunnarFarneback commented:
When is the tuple ordering an interesting property? I would find it very surprising and quite annoying if `a[ind2sub(size(a), k)...]` was pointing to anything other than `a[k]`, especially since that's kind of the idea of `ind2sub` in the first place.

(Personally I'm not too happy about `ind2sub` returning a tuple instead of a vector, since I rarely use `ind2sub` unless I want to do computations on the subscripts, but Julia makes it easy enough to convert between them and as Tim writes `ind2sub` is less needed in Julia than in Matlab.)

tknopp commented:
Agree with @GunnarFarneback on the tuple/array back and forth being not so convenient. But once we have fixed size arrays (in the form of tuples or a small wrapper around that) it would be good to have basic arithmetic on these so that one can stay on the tuple/fixedSizeArray side.

timholy commented:
It's actually much better that it's a tuple, because the compiler can inline on the length of the tuple. If you want to do arithmetic on the "tuples," you probably just want to define this function:
```
myind2sub(dims, ind) = CartesianIndex(ind2sub(sz, i))
```
and you will be a very happy camper :smile:. Elementwise `+`, `-`, `min`, and `max` supported.

garrison commented:
@GunnarFarneback I work in solid state physics, where much of what we do is defined on a lattice.  Often it is necessary to assign some sort of enumeration to the points on a finite-sized lattice; `ind2sub` and `sub2ind` are the natural candidates to map back and forth between indices and lattice site labels.  Oftentimes (such as for a Jordan-Wigner transformation, or for certain sums over all pairs of sites), I want to apply a transformation based on the relative ordering of two sites.  Right now, the two different forms (indices and site labels) give two different orderings for which site comes first.  This is turns out to be incredibly annoying, and means that I must remember to map everything into either indices or site labels before comparing anything.  In the best case this is inefficient; in the worst case (if I map inconsistently) it results in a bug in my code.

I like @timholy's idea of a variant of these functions for row-major ordering.  Looking at the existing uses of `ReverseOrdering`, I don't think it is appropriate here. (I am not trying to define an ordering that is Z-A instead of A-Z; I am trying to get the tuples reversed so that they order differently.)  Perhaps `RowMajor` would be a better name for this flag?  And then `ColumnMajor` could be declared explicitly as well.  This would also work well if row-major arrays (#5932) were to be implemented in some form in the future.

mbauman commented:
I really like the direction this is heading.  This version of `sub2ind` doesn't handle linear indexing in the last index, which is currently supported in base but untested.  The way you've set this up makes it really easy to fix:

```diff
diff --git a/base/abstractarray.jl b/base/abstractarray.jl
index fecf83e..f832f3b 100644
--- a/base/abstractarray.jl
+++ b/base/abstractarray.jl
@@ -1031,7 +1031,9 @@ function repmat(a::AbstractVector, m::Int)
 end

 sub2ind(dims::()) = 1
+sub2ind(dims::(Integer,)) = 1
 sub2ind(dims::(Integer,), i1::Integer) = i1
+sub2ind(dims::(Integer,Integer), i1::Integer) = i1
 sub2ind(dims::(Integer,Integer), i1::Integer, i2::Integer) = i1+dims[1]*(i2-1)
 sub2ind(dims::(Integer,Integer,Integer...), i1::Integer, i2::Integer, I::Integer...) =
     i1 + dims[1]*(sub2ind(tail(dims),i2,I...)-1)
```

+1 to a rebase and merge.

mbauman commented:
It would also be nice to support trailing ones beyond the number of dimensions in `dims`.  Without bounds checking, this could just be `+sum(I)-length(I)`.  Currently, using more indices than dimensions either causes an error or gives an incorrect answer depending on how many indices there are.

Jutho commented:
@mbauman , that's indeed true. So far I mainly focussed on getting `ind2sub` better, but `sub2ind` could certainly improved/generalized along the lines you suggest. I'll take care of it soon.

Jutho commented:
I think this represents the minimal code that cover all cases of both functions
```julia
sub2ind(dims::()) = 1
sub2ind(dims::(),i1::Integer, I::Integer...) = i1==1 ? sub2ind(dims,I...) : throw(BoundsError())
sub2ind(dims::(Integer,Integer...), i1::Integer) = i1
sub2ind(dims::(Integer,Integer...), i1::Integer, I::Integer...) = i1 + dims[1]*(sub2ind(Base.tail(dims),I...)-1)

ind2sub(dims::(), ind::Integer) = ind==1 ? () : throw(BoundsError())
ind2sub(dims::(Integer,), ind::Integer) = (ind,)
@inline function ind2sub(dims::(Integer,Integer...), ind::Integer)
    ind2 = div(ind-1,dims[1])+1
    tuple(ind-dims[1]*(ind2-1), ind2sub(Base.tail(dims),ind2)...)
end
```
With respect to the error behavior: these functions throw an error if otherwise they would return an incorrect result that would not trigger an error when the actual indexing using the returned value takes place. If the result is in principle correct but will subsequently trigger a `BoundsError` when actually indexing the array, these functions don't bother to already throw an error.

Because of the explicit need of `@inline`, these definitions cannot at their current location in `abstractarray.jl` unless some alternative macro is used to add the metadata. Another alternative which could also further improve the performance is to use a `stagedfunction` for the last `ind2sub` definition. Despite previous remarks against that, I think these are not really applicable if #10525 gets merged, since that PR is completely defined in terms of `stagedfunction`s, so one more from `ind2sub` really shouldn't matter.

@mbauman , maybe this PR can be incorporated in your work in #10525. You can then find the most appropriate place to put these definitions such that everything builds successfully?

mbauman commented:
I extended your work like this: https://github.com/JuliaLang/julia/commit/db7f9f5fbe9f2d1e98f33ba9f3153b339058b189.  I've not looked closely at it, but I think that does what we want, no?

romanShekhtman commented:
I am not sure how fast the last implementation is, but a straight forward implementation seems to be much faster. I observed factors of up to 50 

julia> include("test_sub2ind.jl")
Julia sub2ind 2D elapsed time: 3.071918131 seconds
My sub2ind 2D    elapsed time: 0.061298908 seconds
Julia ind2sub 2D elapsed time: 1.626608852 seconds
My ind2sub 2D    elapsed time: 0.253709231 seconds
Julia sub2ind 3D elapsed time: 4.206525844 seconds
My sub2ind 3D    elapsed time: 0.080784239 seconds
Julia ind2sub 3D elapsed time: 2.85505408 seconds
My ind2sub 3D    elapsed time: 0.491045143 seconds


```julia
function mysub2ind( s::Vector{Int64}, 
                    ii::Vector{Int64},jj::Vector{Int64} )
   # 2D sub2ind                 
   const nn = length(ii)
   
   if length(jj) != nn
      error("length(jj) != nn")
   elseif length(s) != 2
      error("length(s) != 2")
   end
   
   idx = Array(Int64,nn)
   
   const m = s[1]
   
   for i = 1:nn
      idx[i] = ii[i] + (jj[i]-1)*m
   end
   
   return idx
end  # function mysub2ind

#----------------------------------------------------

function mysub2ind( s::Vector{Int64},
                    ii::Vector{Int64},jj::Vector{Int64},kk::Vector{Int64} )
   # 3D sub2ind                 
   const nn = length(ii)
   
   if length(jj) != nn || length(kk) != nn
      error("length(jj) != nn")
   elseif length(s) != 3
      error("length(s) != 3")
   end
   
   idx = Array(Int64,nn)
   
   const m = s[1]
   const mn = s[1] * s[2]
   
   for i = 1:nn
      idx[i] = ii[i] + (jj[i]-1)*m + (kk[i]-1)*mn
   end
   
   return idx
end  # function mysub2ind

#-------------------------------------------------------
#-------------------------------------------------------

function myind2sub( s::Vector{Int64}, idx::Vector{Int64} )
   const nn = length(idx)
   
   if length(s) == 2
      ii = Array(Int64,nn)
      jj = Array(Int64,nn)
      
      const s1 = s[1]
   
      for i = 1:nn
      
         it = idx[i]
         ji = div(it - 1, s1) + 1
         jj[i] = ji
      
         ii[i] = it - (ji - 1)*s1
   
      end  # i
   
      return ii,jj
      
   elseif length(s) == 3
   
      ii = Array(Int64,nn)
      jj = Array(Int64,nn)
      kk = Array(Int64,nn)
      
      const s1 = s[1]
      const s12 = s[1] * s[2]
   
      for i = 1:nn
         idxi = idx[i]
         ki = div(idxi - 1, s12) + 1
         kk[i] = ki
      
         it = idxi - (ki - 1)*s12
      
         ji = div(it - 1, s1) + 1
         jj[i] = ji
         ii[i] = it - (ji - 1)*s1
   
      end  # i
   
      return ii,jj,kk
   
   else
      error("myind2sub only works for 2d or 3d arrays.")
   end
   
end # function myind2sub
```

Now the test
```julia

include("mysub2ind.jl")

# 2D test
mrow = 10000
ncol = 20000

pfil = 0.1
nz = trunc(Int64, mrow*ncol*pfil)  # # of non-zeros

ii = rand(1:mrow, nz)
jj = rand(1:ncol, nz)


print("Julia sub2ind 2D ")
tic()
idx = sub2ind( (mrow,ncol), ii,jj );
toc()

print("My sub2ind 2D    ")
tic()
idx2 = mysub2ind( [mrow,ncol], ii,jj );
toc()


er = norm(idx - idx2)
if er != 0
   error("mysub2ind error")
end

print("Julia ind2sub 2D ")

tic()
ii2,jj2 = ind2sub((mrow,ncol), idx)
toc()
er = norm(ii2 - ii) + norm(jj2 - jj)
if er != 0
   error("ind2sub error")
end

print("My ind2sub 2D    ")

tic()
ii3,jj3 = myind2sub([mrow,ncol], idx2)
toc()

er = norm(ii3 - ii) + norm(jj3 - jj)
if er != 0
   error("myind2sub error")
end

#------------------------------------------------

# 3D test

m1 = 1024
m2 = 2048
m3 = 512

pfil = 0.02
nz = trunc(Int64, m1*m2*m3*pfil)  # # of non-zeros

ii = rand(1:m1, nz)
jj = rand(1:m2, nz)
kk = rand(1:m3, nz)

print("Julia sub2ind 3D ")

tic()
idx = sub2ind( (m1,m2,m3), ii,jj,kk );
toc()

print("My sub2ind 3D    ")

tic()
idx2 = mysub2ind( [m1,m2,m3], ii,jj,kk );
toc()

er = norm(idx - idx2)
if er != 0
   error("mysub2ind error")
end

print("Julia ind2sub 3D ")

tic()
ii2,jj2,kk2 = ind2sub((m1,m2,m3), idx)
toc()
er = norm(ii2 - ii) + norm(jj2 - jj) + norm(kk2 - kk)
if er != 0
   error("ind2sub error")
end

print("My ind2sub 3D    ")

tic()
ii3,jj3,kk3 = myind2sub([m1,m2,m3], idx2)
toc()

er = norm(ii3 - ii) + norm(jj3 - jj) + norm(kk3 - kk)
if er != 0
   error("myind2sub error")
end

```

StefanKarpinski commented:
> `error("myind2sub only works for 2d or 3d arrays.")`

This is kind of a crucial limitation.

romanShekhtman commented:
Yes, I agree that this is a crucial limitation.  The point of my post is to show that it is possible to write sub2ind and ind2sub that are significantly faster than Julia's implementation.

ehaber99 commented:
I see the point of wanting to have a generic code but when this generic code is 50 times slower than a simple implementation then I think that we should ask ourselves if having a generic code is useful. In our work, 2 and 3D suffice. We can certainly use the multiple dispatch to call our functions when we are using it in 2 or 3D but I dare to say that it is probably easy to generate a less elegant but way faster implementation for higher dimensions

StefanKarpinski commented:
I'm not sure who you're arguing against. Everyone wants the fast version.

mbauman commented:
Would you be able to benchmark your tests against this brach?  I'd like to see this merged as well, and this PR presents a significant improvement over the status quo.

Jutho commented:
Note that this PR so far did not yet include any changes of the vectorized versions of `sub2ind` and `ind2sub`, which is what the benchmark of @romanShekhtman is about. But I guess there is indeed some room for improvement there as well.

StefanKarpinski commented:
@JeffBezanson, it seems like your objections to using staged functions may have stalled the merging of this branch. Since it no longer uses staged functions and is much faster than the current version, are you ok with merging it?

Jutho commented:
I guess the other part of the reason for not (yet) merging this is that it will also be included in @mbauman 's PR on more generic indexing?

mbauman commented:
All the more reason **to** merge it!  I'll happily rebase over this.

StefanKarpinski commented:
Yes, it's generally better to merge changes in stages than all at once. That way if something goes wrong, it's easier to localize what caused the problem.

mbauman commented:
With respect to vectorized ind2sub, I think the path forward there is to use @simonster's fast dividing integers with the existing implementation.  If I remember right @timholy has that implementation in his ReshapedArrays (#10507).

mbauman commented:
I think the other missing piece here was the fact that you can't use the `@inline` macro this early in the bootstrap, but you can now make use of the internal `@_inline_meta` macro to work around this; see #10704 for details.

timholy commented:
I too vote for merging this.

@mbauman you are correct. @simonster's fast dividing integers make it about a factor of 3 better than calling `div`, _if_ you don't worry about the cost of computing the magic numbers in the first place.

Jutho commented:
I am certainly also in favor of merging, but now I remember why I was holding of. The best timings used a definition of `ind2sub` that required an explicit `@inline`. But since that is not yet available in `abstractarray.jl` using bootstrap, this needed the workaround `@_inline_meta` also implemented by @mbauman in the context of #10525 .

StefanKarpinski commented:
How about merging the version with less-than-perfect-but-better timings now and opening an issue to track the better timing for a future change?

mbauman commented:
@Jutho - that's already in master: https://github.com/JuliaLang/julia/blob/22a73f914ab79c7dd8293e498bc16e536f305510/base/abstractarray.jl#L343-L349

Jutho commented:
Ah I missed that. Great. Let me first change that and take a final look before merging.

Jutho commented:
I also now notice that you made the same comment about `@_inline_meta` 3 minutes before me. My computer crashed somewhere in the mean time so I probably missed that.

Jutho commented:
I rebased and included the final method definitions reported [here](https://github.com/JuliaLang/julia/pull/10337#issuecomment-81251164). 

Following the suggestion of @timholy  [here](https://github.com/JuliaLang/julia/pull/10337#issuecomment-78189849) I have removed all definitions where `dims` was specified as a vector instead of a tuple. I have added a version where an array as first argument means it just calls the same method using `size(a)`. This definition was already there for `ind2sub` but not `sub2ind`.

Finally, I modified `ind2sub!` to a much shorter and faster (not benchmarked, but there are many less `div` instructions) version.

I hade some git struggles but don't find how to kill the CIs for the first two commits, everything is in the last commit. I am actually surprised that there are still three different commits being shown, I thought I squashed them. Also, I am currently unable to build julia locally because if [this](http://stackoverflow.com/questions/29529455/missing-c-header-debug-after-updating-osx-command-line-tools-6-3) so this is untested and will still contain an error somewhere with high probability.

StefanKarpinski commented:
I'm only seeing one commit, so maybe you need to hit refresh or something. Let's see what the CI says and if that passes, I think it's safe to merge this.

mbauman commented on base/abstractarray.jl:
```diff
@@ -984,69 +984,26 @@ function repmat(a::AbstractVector, m::Int)
-    return index
-end
 
-function sub2ind{T<:Integer}(dims::Array{T}, sub::Array{T})
```
I really want to use the new definition for this, but I'm afraid that this meaning should be deprecated first.

Jutho commented:
Ah yes, I should have taken care of that. I will try to fix this tonight.

Jutho commented:
I am a bit stuck with the deprecation:
* `ind2sub(a::AbstractArray,ind::Integer)` is currently already defined to call `ind2sub` with `size(a)`.
* `sub2ind` is defined in various ways such that the first argument `dims` can be represented as an array instead of a tuple. 

So I cannot introduce a `ind2sub(a::AbstractArray,ind::Integer) = ind2sub(size(a),ind)` method without this being ambiguous with the deprecated `sub2ind` methods where `dims` is an array. However, not introducing `ind2sub(a::AbstractArray,ind::Integer)` completely breaks the symmetry with `sub2ind`, unless I also get rid (via a deprecation) of the `sub2ind(a::AbstractArray,...)` definition.

Then these methods only work with a `dims::(Integer...)` and one has to manually call `size(a)`; maybe that is not too bad?

timholy commented:
Given that `ind2sub` already has the behavior we eventually want, I think we should live with the asymmetry. My vote is to deprecate `sub2ind(::AbstractArray, sub)` in 0.4, and then introduce the new method in 0.5.

ViralBShah commented:
Even though it is not clear, it would be great if we can find a way to fix the APIs here in 0.4. It would be nice to end up with as stable an array API as we can in 0.4.

Jutho commented:
I corrected the above comment since it incorrectly switched `ind2sub` and `sub2ind` on various places. 

The only consistent definition for a stable array API is probably to deprecate any use of `AbstractArray` as first argument, either as dims or as `size(a)`. The latter use can then be reintroduced later.

ViralBShah commented:
That seems reasonable.

Jutho committed:
ind2sub / sub2ind redesign

Jutho commented:
Rebased and updated for the post-tuple-redesign era.

I dediced to keep `ind2sub(a::AbstractArray,ind::Integer)` since there seems to be an explicit PR ( #9256 ) and request ( #294 ) for it. Not sure why it was not decided to make this symmetric back then, but as @timholy said, we will have to live with this until v0.5. Or we make this a breaking change right now?

mbauman commented:
There are some strange performance effects that I'm seeing here.

On my #10525 branch with some similar definitions (https://github.com/JuliaLang/julia/commit/abba59d57d31e65ce6a0e91702b1bff77d899dc1), I get nice behavior for 2-dimensional sub2ind:

```llvm
julia> @code_llvm sub2ind((2,2), 2, 2)

define i64 @julia_sub2ind_683679([2 x i64], i64, i64) {
top:
  %3 = extractvalue [2 x i64] %0, 0
  %4 = add i64 %2, -1
  %5 = mul i64 %4, %3
  %6 = add i64 %5, %1
  ret i64 %6
}
```

But here:

```llvm
julia> @code_llvm sub2ind((2,2), 2, 2)

define %jl_value_t* @julia_sub2ind_43908(%jl_value_t*, %jl_value_t**, i32) {
top:
  %3 = alloca [7 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [7 x %jl_value_t*]* %3, i64 0, i64 0
  %4 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 2
  store %jl_value_t* inttoptr (i64 10 to %jl_value_t*), %jl_value_t** %.sub, align 8
  %5 = load %jl_value_t*** @jl_pgcstack, align 8
  %6 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 1
  %.c = bitcast %jl_value_t** %5 to %jl_value_t*
  store %jl_value_t* %.c, %jl_value_t** %6, align 8
  store %jl_value_t** %.sub, %jl_value_t*** @jl_pgcstack, align 8
  %7 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 3
  store %jl_value_t* null, %jl_value_t** %7, align 8
  %8 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 4
  store %jl_value_t* null, %jl_value_t** %8, align 8
  %9 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 5
  store %jl_value_t* null, %jl_value_t** %9, align 8
  %10 = getelementptr [7 x %jl_value_t*]* %3, i64 0, i64 6
  store %jl_value_t* null, %jl_value_t** %10, align 8
  store %jl_value_t* null, %jl_value_t** %4, align 8
  %11 = load %jl_value_t** %1, align 8
  %12 = getelementptr %jl_value_t** %1, i64 1
  %13 = load %jl_value_t** %12, align 8
  %14 = bitcast %jl_value_t* %13 to i64*
  %15 = load i64* %14, align 8
  %16 = getelementptr %jl_value_t** %1, i64 2
  %17 = add i32 %2, -2
  %18 = call %jl_value_t* @jl_f_tuple(%jl_value_t* null, %jl_value_t** %16, i32 %17)
  store %jl_value_t* %18, %jl_value_t** %4, align 8
  %19 = load %jl_value_t** inttoptr (i64 4389925360 to %jl_value_t**), align 16
  store %jl_value_t* %19, %jl_value_t** %7, align 8
  %20 = load %jl_value_t** inttoptr (i64 4385340336 to %jl_value_t**), align 16
  store %jl_value_t* %20, %jl_value_t** %8, align 8
  %21 = getelementptr %jl_value_t* %11, i64 1
  %22 = bitcast %jl_value_t* %21 to i64*
  %23 = load i64* %22, align 8
  %24 = insertvalue [1 x i64] undef, i64 %23, 0
  %25 = insertvalue [1 x [1 x i64]] undef, [1 x i64] %24, 0
  %26 = call %jl_value_t* @allocobj(i64 8)
  %27 = getelementptr inbounds %jl_value_t* %26, i64 -2, i32 0
  store %jl_value_t* inttoptr (i64 4385798496 to %jl_value_t*), %jl_value_t** %27, align 8
  %28 = bitcast %jl_value_t* %26 to [1 x [1 x i64]]*
  store [1 x [1 x i64]] %25, [1 x [1 x i64]]* %28, align 8
  store %jl_value_t* %26, %jl_value_t** %9, align 8
  %29 = load %jl_value_t** %4, align 8
  store %jl_value_t* %29, %jl_value_t** %10, align 8
  %30 = call %jl_value_t* @jl_f_apply(%jl_value_t* null, %jl_value_t** %7, i32 4)
  %31 = bitcast %jl_value_t* %30 to i64*
  %32 = load i64* %31, align 8
  %33 = bitcast %jl_value_t* %11 to i64*
  %34 = load i64* %33, align 8
  %35 = add i64 %32, -1
  %36 = mul i64 %34, %35
  %37 = add i64 %36, %15
  %38 = call %jl_value_t* @jl_box_int64(i64 signext %37)
  %39 = load %jl_value_t** %6, align 8
  %40 = getelementptr inbounds %jl_value_t* %39, i64 0, i32 0
  store %jl_value_t** %40, %jl_value_t*** @jl_pgcstack, align 8
  ret %jl_value_t* %38
}
```

Both branches have the GC frame and boxing for 3-dimensional sub2ind.  I don't understand why this is occurring (and I don't read LLVM well enough to even understand what it's doing), but it'd be really nice if we could get these definitions to optimize better.  It's almost essential for #10525.

mbauman commented:
Aha! Since these are inlined, a function such as `f(sz, i1, i2, i3) = sub2ind(sz, i1, i2, i3)` removes all overhead.  I can't seem to get that to work from a generated function, though.  Is this something that could conceivably work?

```jl
julia> @generated function f(sz,I...)
          :(sub2ind(sz, I...))
       end
f (generic function with 1 method)

julia> @code_llvm f((2,2,2),2,2,2) # Long

julia> @generated function g(sz,I...)
          idxs = Expr[:(I[$d]) for d=1:length(I)]
          :(sub2ind(sz, $(idxs...)))
       end
g (generic function with 1 method)

julia> @code_llvm g((2,2,2),2,2,2) # Long

julia> @generated function h(sz,I...)
          N = length(I)
          quote
             @Base.Cartesian.nexprs $N d->(I_d = I[d])
             @Base.Cartesian.ncall $N sub2ind sz I
          end
       end
h (generic function with 1 method)

julia> @code_llvm h((2,2,2),2,2,2) # Long
```

Jutho commented:
This branch is certainly not ready to be merged, I also noticed it had a detrimental effect on the vectorized versions of `ind2sub` and `sub2ind`, i.e. the code of @romanShekhtman became much worse. I will try to spend some more time looking into this tonight, but the only way I know to really avoid any issues is to use the `@generated` definitions that I had at the very beginning of this PR. Since these functions are not used very common throughout base and the few methods that do rely on them are also all of the `@generated` type, I don't really see the harm in this. But that's not up to me to decide ...

Jutho commented:
Forcing `sub2ind` to be inlined seems to fix it also for `@generated` functions. I know use the following definitions:
```julia
sub2ind(dims::Tuple{}) = 1
sub2ind(dims::Tuple{},I::Integer...) = +(I...) - length(I) + 1
sub2ind(dims::Tuple{Integer,Vararg{Integer}}, i1::Integer) = i1
function sub2ind(dims::Tuple{Integer,Vararg{Integer}}, i1::Integer, I::Integer...)
    @_inline_meta()
    i1 + dims[1]*(sub2ind(tail(dims),I...)-1)
end
```
I suspect the function which marks a method as inlineable is not run on the code generated of a generated function, but I have no idea whatsoever about what's going on in `inference.jl`. 

I have to run now, but will update this PR tonight.

ehaber99 commented:
I must admit that I have a hard time to understand why this issue has not been resolved.
The following is a simple modification by @romanShekhtman that works for any number of dimensions


``` Julia

#import Base.sub2ind
function mysub2ind( st::Tuple{Integer,Vararg{Integer}},  # s::Vector{Int64},
                    ii::Vector{Int64}... )
   
   const ndim = length(st)

   # Copy from tuple to vector, otherwise the
   # code will run very slowly.
   s = Array{Int64}(ndim)
   for j = 1:ndim
      s[j] = st[j]
   end

   if ndim < 2
      error("ndim < 2")
   elseif length(ii) != ndim
      error("length(ii) != ndim")
   end

   const nn = length(ii[1])

   for j = 2:ndim
      if length(ii[j]) != nn
         error("length(ii[j]) != nn")
      end
   end  # j
   
   
   idx = copy(ii[1])
  
   m = s[1]
   
   for j = 2:ndim
      iij = ii[j]
      for i = 1:nn
         idx[i] += (iij[i] - 1) * m 
      end
      
      m *= s[j]
   end  # j
  
   return idx
end  # function mysub2ind

#-------------------------------------------------------

#import Base.ind2sub
function myind2sub( st::Tuple{Integer,Vararg{Integer}}, #s::Vector{Int64},  #s::(Int64,Int64...), # 
                    idx::Vector{Int64} )

   const nn = length(idx)
   const ndim = length(st)
   if ndim < 2
      error("ndim < 2")
   end

   # Copy from tuple to vector, otherwise the
   # code will run very slowly.
   s = Array{Int64}(ndim)
   for j = 1:ndim
      s[j] = st[j]
   end


   ii = cell(ndim)

   it = copy(idx)
   m = prod(s[1:ndim-1])
   id = 0

   for j = ndim:-1:2
      iij = Array{Int64}(nn)

      for i = 1:nn
         id = div(it[i]-1, m) + 1
         iij[i] = id
         it[i] -= (id-1)*m
      end  # i

      m = div(m, s[j-1])
      ii[j] = iij
   end  # j
   
   ii[1] = it

   return ii
end # function myind2sub

```

The code is tested  in

``` Julia
include("mysub2indN.jl")

# 2D test
mrow = 10000
ncol = 20000

pfil = 0.1
nz = trunc(Int64, mrow*ncol*pfil)  # # of non-zeros

ii = rand(1:mrow, nz)
jj = rand(1:ncol, nz)


print("Julia sub2ind 2D ")
tic()
idx = sub2ind( (mrow,ncol), ii,jj );
toc()

print("My sub2ind 2D    ")
tic()
idx2 = mysub2ind( (mrow,ncol), ii,jj );
toc()


er = norm(idx - idx2)
if er != 0
   error("mysub2ind error")
end

print("Julia ind2sub 2D ")

tic()
ii2,jj2 = ind2sub((mrow,ncol), idx)
toc()
er = norm(ii2 - ii) + norm(jj2 - jj)
if er != 0
   error("ind2sub error")
end

print("My ind2sub 2D    ")

tic()
ii3,jj3 = myind2sub((mrow,ncol), idx2)
#ii3,jj3 = myind2sub([mrow,ncol], idx2)
toc()

er = norm(ii3 - ii) + norm(jj3 - jj)
if er != 0
   error("myind2sub error")
end

ii=0; jj=0; ii2=0; jj2=0; idx=0; idx2=0;

#------------------------------------------------

# 3D test

m1 = 1024
m2 = 2048
m3 = 512
dim = (m1,m2,m3)

pfil = 0.02
nz = trunc(Int64, m1*m2*m3*pfil)  # # of non-zeros

ii = rand(1:m1, nz)
jj = rand(1:m2, nz)
kk = rand(1:m3, nz)

print("Julia sub2ind 3D ")

tic()
idx = sub2ind( dim, ii,jj,kk );
toc()

print("My sub2ind 3D    ")

tic()
idx2 = mysub2ind( dim, ii,jj,kk );
toc()

er = norm(idx - idx2)
if er != 0
   error("mysub2ind error")
end

print("Julia ind2sub 3D ")

tic()
ii2,jj2,kk2 = ind2sub(dim, idx)
toc()
er = norm(ii2 - ii) + norm(jj2 - jj) + norm(kk2 - kk)
if er != 0
   error("ind2sub error")
end

print("My ind2sub 3D    ")

tic()
ii3,jj3,kk3 = myind2sub(dim, idx2)
#ii3,jj3,kk3 = myind2sub([m1,m2,m3], idx2)
toc()

er = norm(ii3 - ii) + norm(jj3 - jj) + norm(kk3 - kk)
if er != 0
   error("myind2sub error")
end

ii=0; jj=0; kk=0; ii2=0; jj2=0; kk2=0; ii3=0; jj3=0; kk3=0;
idx=0; idx2=0;

#------------------------------------------------

# 5D test

m = [ 100 500 200 300 600 ]

pfil = 0.000001
nz = trunc(Int64, prod(m)*pfil)  # # of non-zeros

ii = cell(5)
ii[1] = rand(1:m[1], nz)
ii[2] = rand(1:m[2], nz)
ii[3] = rand(1:m[3], nz)
ii[4] = rand(1:m[4], nz)
ii[5] = rand(1:m[5], nz)
dim = (m[1],m[2],m[3],m[4],m[5])

print("Julia sub2ind 5D ")

tic()
idx = sub2ind( dim, ii[1],ii[2],ii[3],ii[4],ii[5] );
toc()

print("My sub2ind 5D    ")

tic()
idx2 = mysub2ind( dim, ii[1],ii[2],ii[3],ii[4],ii[5] );
toc()

er = norm(idx - idx2)
if er != 0
   error("mysub2ind error")
end

print("Julia ind2sub 5D ")

ii2 = cell(5)
tic()
ii2[1],ii2[2],ii2[3],ii2[4],ii2[5] = ind2sub(dim, idx)
toc()

er = 0
for k=1:5
   er += norm(ii2[k] - ii[k])
end
if er != 0
   error("ind2sub error")
end


print("My ind2sub 5D    ")
ii3 = cell(5)

tic()
ii3[1],ii3[2],ii3[3],ii3[4],ii3[5] = myind2sub(dim, idx2)
toc()

er = 0
for k=1:5
   er += norm(ii3[k] - ii[k])
end
if er != 0
   error("myind2sub error")
end

```

The results are

```
Julia sub2ind 2D elapsed time: 2.029261623 seconds
My sub2ind 2D    elapsed time: 0.070300548 seconds
Julia ind2sub 2D elapsed time: 1.779486347 seconds
My ind2sub 2D    elapsed time: 0.25167268 seconds
Julia sub2ind 3D elapsed time: 2.699629905 seconds
My sub2ind 3D    elapsed time: 0.107105745 seconds
Julia ind2sub 3D elapsed time: 3.22423048 seconds
My ind2sub 3D    elapsed time: 0.50075644 seconds
Julia sub2ind 5D elapsed time: 2.414871148 seconds
My sub2ind 5D    elapsed time: 0.026106956 seconds
Julia ind2sub 5D elapsed time: 1.943472252 seconds
My ind2sub 5D    elapsed time: 0.090888286 seconds
```

What strikes me is that @romanShekhtman implementation is the straight way to do this and it gives the best timing. Maybe it would be wise to adopt his code (I have already done so) and not to rewrite it

timholy commented:
@ehaber99, the versions that @Jutho and @mbauman are working on will be even better, especially for the case that most of us are concerned about: converting a single `sub` to a single `Int` (which will lie at the core of #10525). The performance differences could easily be a factor of 10 (maybe more): it's pretty hard to argue that a solution that starts up a bunch of loops, allocates memory, etc. is better than one that is just a few clock ticks (a couple of adds and multiplies). So no one is wasting their time by putting good effort into this.

@mbauman and @Jutho, yeah, the splatting is a big, big problem. Given that we already use generated function elsewhere in base for the sole purpose of avoiding splatting, I decree (for what it's worth): go with the generated function version. Once the performance hit from splatting has been resolved, we can comb through base and rewrite things. (Maybe post your code to a gist first, though!)

Jutho commented:
@ehaber99 , the topic of this PR was not to improve the vectorized version of `ind2sub` and its counterpart, but just the versions for a single value of the arguments. I agree that also the vectorized version needs to be improved, but it was decided a few comments above that that would be the topic of a second PR. Currently, the vectorized form just calls the single value form in combination with some `map` construct. 

The single value methods are certainly already performing much better with the definitions in the current commit. Essential to this is that the recursive definitions of `ind2sub` and `sub2ind` are getting inlined; the latter has an explicit `@_inline_meta()` call, the former did not. Remaining issues:
* With the current commit, the definition of `ind2sub` does not seem to get inlined when called from within a `@generated` function. As discussed in my previous post, this seems to be fixed by also adding an explicit `@_inline_meta()` construct in that definition. Maybe `inline_worthy` is not called when compiling the expression body of a `@generated` function, but that's just a wild guess?
* The current definitions of the single argument `ind2sub` and `sub2ind` seem to combine less well with the `map` construction currently used in the definition of the vectorized methods. I have no idea why this is the case. So maybe we need to replace the definition of the vectorized functions straight away within this PR after all. I see two possibilities:
1) The vectorized version still calls the single value methods `ind2sub` and `sub2ind`, but to get rid of any type uncertainty the vectorized method itself uses a `@generated` definition.
2) Along the lines of the proposal of @ehaber99 , which might even be more cache friendly or susceptible to a simd speedup

Jutho commented:
I finally understand what @mbauman and probably @timholy were talking about. It's not the inlining of `ind2sub` that goes wrong, it's just that the `@generated` functions still have a vararg argument which is causing all the overhead, even though dedicated code is generated for every given argument length.

timholy commented:
But don't be confused about what you see in a "naked" call from the REPL and how it will work out in practice when compiled into functions. If you copy the example in http://docs.julialang.org/en/latest/manual/metaprogramming/#an-advanced-example, then
```jl
g(dims, i1, i2, i3, i4, i5, i6) = sub2ind_gen(dims, i1, i2, i3, i4, i5, i6)
g (generic function with 1 method)

julia> @code_llvm g((2,2,2,2,2,2), 2, 2, 2, 2, 2, 2)

define i64 @julia_g_44425([6 x i64], i64, i64, i64, i64, i64, i64) {
top:
  %7 = extractvalue [6 x i64] %0, 0
  %8 = add i64 %2, -1
  %9 = extractvalue [6 x i64] %0, 1
  %10 = add i64 %3, -1
  %11 = extractvalue [6 x i64] %0, 2
  %12 = add i64 %4, -1
  %13 = extractvalue [6 x i64] %0, 3
  %14 = add i64 %5, -1
  %15 = extractvalue [6 x i64] %0, 4
  %16 = add i64 %6, -1
  %17 = mul i64 %16, %15
  %18 = add i64 %14, %17
  %19 = mul i64 %18, %13
  %20 = add i64 %12, %19
  %21 = mul i64 %20, %11
  %22 = add i64 %10, %21
  %23 = mul i64 %22, %9
  %24 = add i64 %8, %23
  %25 = mul i64 %24, %7
  %26 = add i64 %25, %1
  ret i64 %26
}
```
which is just gorgeous. In contrast, even if you put `@inline` in front of all the definitions in this PR, you don't get anything nearly so nice. This is why I think that, despite Jeff's reservations, we basically need to go with the generated function version until this gets fixed.

Jutho commented:
Yes that's exactly the point also for the post of @mbauman above. The reason that his normal function gave short LLVM code while the `@generated` ones did not was just because the short one had no more varargs whereas all of the different `@generated` versions were still using varargs. Further using those in another function which specifies the number of arguments also yields short code; that's what got me confused above. But most importantly, this is true with the current definitions of `ind2sub` and `sub2ind`, so these can remain as they are.

To make the vectorized definitions of these functions fast, I needed to implement new code along the lines suggested by @ehaber99 . I tweaked it some more and now obtain for this example, with the commit that I will push in a minute:
```
Julia sub2ind 2D elapsed time: 0.10945286 seconds
My sub2ind 2D    elapsed time: 0.12169844 seconds
Julia ind2sub 2D elapsed time: 0.355919331 seconds
My ind2sub 2D    elapsed time: 0.423536101 seconds
Julia sub2ind 3D elapsed time: 0.151631702 seconds
My sub2ind 3D    elapsed time: 0.165676097 seconds
Julia ind2sub 3D elapsed time: 0.704246112 seconds
My ind2sub 3D    elapsed time: 0.696756499 seconds
Julia sub2ind 5D elapsed time: 0.015101847 seconds
My sub2ind 5D    elapsed time: 0.031233035 seconds
Julia ind2sub 5D elapsed time: 0.118229111 seconds
My ind2sub 5D    elapsed time: 0.142487001 seconds
```
which looks pretty good to me.

timholy commented:
I wonder if I'm misunderstanding you, but my impression is that we're saying opposite things. My point is that the best combination is to implement `sub2ind` as a generated function, and then let it be called by anything (including an ordinary function). You seem to be arguing the reverse. In my example above, `sub2ind_gen` is a generated function, and it gives great results when called from `g`. Unfortunately, the version in this PR gives much worse results.

timholy commented:
Am I misunderstanding your timings? In every case the ones labeled "My" are slower than the ones labeled "Julia."

Jutho commented:
> Am I misunderstanding your timings? In every case the ones labeled "My" are slower than the ones labeled "Julia."

Well, the "my" timings are the one of @ehaber99 's code, the Julia timings are the ones of my new implementation in the latest commit. But apparently there is still an error; I got to tired yesterday evening.

> I wonder if I'm misunderstanding you, but my impression is that we're saying opposite things. My point is that the best combination is to implement sub2ind as a generated function, and then let it be called by anything (including an ordinary function). You seem to be arguing the reverse. In my example above, sub2ind_gen is a generated function, and it gives great results when called from g. Unfortunately, the version in this PR gives much worse results.

Why do you think that. The code generated by the current definitions of `sub2ind` and `ind2sub` is also as compact and fast as I think are possible, but maybe I am overseeing something?

timholy commented:
If there's something small we can change to fix the current ones, we should do that. Maybe it would be worth inserting a breakpoint [here](https://github.com/JuliaLang/julia/blob/dea3d0e42029af05f58a0069491238462382e591/src/codegen.cpp#L1440) to see what's happening.

timholy commented:
> Well, I just run the test file of @ehaber99 against my latest comment, so the "my" timings refer to the code of @ehaber99, the Julia timings are the ones corresponding to my new implementation in the latest commit. 

That makes much more sense, thanks for explaining!

Jutho commented:
Ok, so that's the final, test-succeeding commit with does not used `@generated` functions. I can now replace them with the `@generated` definitions, but it might be useful to keep that as two separate commits for further reference.

Jutho committed:
generated ind2sub / sub2ind definitions

Jutho commented:
I think this is more or less ready. I guess it could use a final review; it ain't much, just a few lines of code.
 
The only point I am not fully satisfied with is about the special casing of `ind2sub` for `N=0`. The general policy we decided on was not to throw `BoundsError`s from `ind2sub` or `sub2ind` but just produce a consistent result (which would also be out of bounds and thus trigger an error when used further down to actually index an array). This is however impossible with `ind2sub` for a 0-dimensional array, where the only result can be the empty tuple `()`. I do not see a way around this.

Jutho commented:
One more fix after all; with this I guess the output of all `ind2sub` and `sub2ind` methods can be completely inferred.

timholy commented on base/abstractarray.jl:
```diff
@@ -1007,97 +1007,79 @@ function repmat(a::AbstractVector, m::Int)
+        end
     end
-    return index
+    return :($ex + 1)
```
Does it help to `Expr(:meta, :inline)` this to avoid the allocation of the creation of the tuple? Or does this always get inlined?

Jutho replied:
It probably does not get inlined. Would it be necessary for `sub2ind` to be inlined? Before this was necessary because of the recursive definition. Probably it is still necessary if this is relied on in indexing? I will restore the exlicit inline behavior.

timholy commented:
LGTM. 

I don't know if it's a coincidence, but that travis segfault happened in arrayops, which is where `ind2sub` is tested. It's therefore possible that this is revealing a deeper problem, but since this also passes the tests frequently  I wouldn't let that possibility hold this up.

It's interesting that we don't seem to test `sub2ind`. #10525 will add a lot of indirect tests for it, however.

Jutho commented:
It seems to be a 32 bit issue, but it is likely caused by this final fix. I will rebase and change the git history such that the final fix is part of the first commit, and add the inline statement discussed in the line comments.

Jutho commented:
All looks good now.

timholy commented:
Thanks for doing this!
